examples/computer_vision/byol_pytorch/model_def.py

```diff
@@ -0,0 +1,367 @@
+)
+from reducers import AvgReducer, ValidatedAccuracyReducer
+
+TorchData = Union[Dict[str, torch.Tensor], Sequence[torch.Tensor], torch.Tensor]
+
+
+def set_learning_rate_warmup_cosine_anneal(
+    hparams: AttrDict,
+    optimizer: Optimizer,
+    global_batch_size: int,
+    batch_idx: int,
+    batches_per_epoch: int,
+) -> None:
+    """
+    Cosine annealing with warmup, as described in BYOL paper.
+    """
+    # Learning rate scales linearly with batch size, is equal to base when global_batch_size==base_batch_size.
+    p = hparams.self_supervised.learning_rate
+    assert hparams.total_epochs > p.warmup_epochs
+    base_lr = p.base * (global_batch_size / p.base_batch_size)
+    fractional_epoch = batch_idx / batches_per_epoch
+    if fractional_epoch <= p.warmup_epochs:
+        # Warmup domain.
+        adjusted_lr = base_lr * (fractional_epoch / p.warmup_epochs)
+    else:
+        # Cosine annealing domain.
+        cosine_progress = (fractional_epoch - p.warmup_epochs) / (
+            hparams.total_epochs - p.warmup_epochs
+        )
+        cosine_multiplier = 1 + math.cos(math.pi * cosine_progress)
+        adjusted_lr = base_lr * cosine_multiplier
+    for param_group in optimizer.param_groups:
+        param_group["lr"] = adjusted_lr
+
+
+def set_ema_beta_cosine_anneal(
+    hparams: AttrDict, byol_model: BYOL, batch_idx: int, batches_per_epoch: int
+) -> None:
+    """
+    Anneals the exponential moving average coefficient, as described in the BYOL paper.
+    """
+    fractional_epoch = batch_idx / batches_per_epoch
+    # 1 − (1 − τbase) · (cos(πk/K) + 1)/2
+    progress = fractional_epoch / hparams.total_epochs
+    ema_beta = 1 - (
+        (1 - hparams.self_supervised.moving_average_decay_base)
+        * (math.cos(math.pi * progress) + 1)
+        / 2
+    )
+    byol_model.target_ema_updater.beta = ema_beta
+
+
+def classifier_loss(
+    hparams: AttrDict, input_logits: torch.Tensor, target: torch.Tensor
+) -> torch.Tensor:
+    """
+    Classifier loss is as described in section C.1 of BYOL paper.
+    """
+    if hparams.classifier.logit_clipping.enabled:
+        clipped_logits = hparams.classifier.logit_clipping.alpha * torch.tanh(
+            input_logits / hparams.classifier.logit_clipping.alpha
+        )
+    else:
+        clipped_logits = input_logits
+    return (
+        F.cross_entropy(clipped_logits, target)
+        + hparams.classifier.logit_regularization_beta * (clipped_logits ** 2).mean()
+    )
+
+
+class BYOLTrial(PyTorchTrial):
+    def __init__(self, context: PyTorchTrialContext) -> None:
+        self.context = context
+        self.hparams = AttrDict(self.context.get_hparams())
+        self.data_config = AttrDict(self.context.get_data_config())
+        self.hparams.total_epochs = self.context.get_experiment_config()["searcher"][
+            "max_length"
+        ]["epochs"]
+        self.rank = self.context.distributed.get_rank()
+        if self.data_config.use_rank_for_download_dir:
```

Comment:
i like using a filelock more since it will only download once per instance instead of once per slot:
https://github.com/determined-ai/determined/blob/master/examples/computer_vision/detr_coco_pytorch/model_def.py#L98

Reply:
This is awesome, wasn't aware of this library!  Done.